cpp/src/arrow/util/bit_util.h

```diff
@@ -322,6 +322,31 @@ static inline uint16_t ByteSwap(uint16_t value) {
   return static_cast<uint16_t>(ByteSwap(static_cast<int16_t>(value)));
 }
 static inline uint8_t ByteSwap(uint8_t value) { return value; }
+static inline int8_t ByteSwap(int8_t value) { return value; }
+static inline double ByteSwap(double value) {
+#if defined(__GNUC__)
+  // avoid dereferencing type-punned pointer will break strict-aliasing rules
```

Comment:
Instead, you should use the "Safe" functions from `util/ubsan.h` on all compilers. For example:
```c++
static inline double ByteSwap(double value) {
  const uint64_t swapped = ARROW_BYTE_SWAP64(SafeCopy<uint64_t>(value));
  return SafeCopy<double>(swapped);
}
```